xnmt/optimizer.py

```diff
@@ -48,3 +48,19 @@ class AdamTrainer(XnmtOptimizer, Serializable):
   def __init__(self, yaml_context, alpha = 0.001, beta_1 = 0.9, beta_2 = 0.999, eps = 1e-8):
     self.optimizer = dy.AdamTrainer(yaml_context.dynet_param_collection.param_col, 
                                     alpha, beta_1, beta_2, eps)
+
+class LazySGDTrainer(XnmtOptimizer, Serializable):
```

Comment:
Is this limited to SimpleSGDTrainer only? It'd be nice if we could do some `args`+`kwargs` magic to make it support any trainer that we want. This could be a future commit though.

Reply:
This is only needed for the default trainer which is currently SGD. The main task can use any trainer it wants, and the auxiliary tasks use the lazy trainer which is not instantiated because only the main task's trainer is used. Not sure if this answers your question/suggestion?

Reply:
Actually, I'm confused. Why don't we use trainers for the auxiliary tasks? Their parameters need to be updated as well, right?

Reply:
Actually, all tasks' parameters are in the same parameter collection, which is updated with the same trainer. So one trainer updates the parameters of all tasks. I'm not sure if it's possible to have individual trainers for different parameter sub-collections, but even if it is that would probably confusing when we start sharing parameters across tasks?

Reply:
It is possible to have different trainers for different sub-collections (I think). I'm not super-happy with a design where we basically introduce a class that doesn't do anything as a band-aid to prevent our program from crashing. Do you think it'd be worth changing where the trainer lives in the code base so we don't have to do this patch? Perhaps make it a property of MultiTaskTrainingRegimen, in which case this trainer would be used by all sub-TrainingRegimens?

Reply:
Yeah, I agree this is a bit hacky. The problem is that SimpleTrainingRegimen also have trainers that are instantiated to the default if set to None, and they know nothing about whether they live inside a multi-task training regimen or not. A few ways to solve this might be:
- pull trainer out of the "train" part and into the "experiments" part, although that seems a bit counter-intuitive
- make the SGDTrainer lazy. That way we have a bit of band-aid code in the SGDTrainer, but at least no extra class definition.
- (maybe my favorite) check during deserialization that only one trainer is instantiated and create a reference to this trainer if more trainers are specified. This can probably be done by using the _xnmt_id mechanism also used for parameter sharing.

What do you think?

Reply:
Hmm, this is also not the absolute best solution, but how about if SimpleTrainingRegimen's trainer is None we don't instantiate it in the constructor, but instead wait until the trainer is used. This is the same thing that LazySGDTrainer is doing, but it seems to make more sense to me and doesn't require the extra class definition or adding too much complexity.

Reply:
I arrived at a solution now by breaking apart SimpleTrainingRegimen into two classes:
- SimpleTrainingTask has no trainer and is used inside the multi-task regimens
- SimpleTrainingRegimen extends it by holding a trainer and specifying the training loop
This allows keeping only one trainer that lives in the top of the hierarchy, as you suggested first.